File: Customer Data Analysis and Insights Project/scripts/python/statistical_analysis.py
"""
Statistical Analysis Script
Descriptive, Inferential, and Exploratory Statistical Analysis
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from scipy.stats import chi2_contingency, normaltest
import warnings
from pathlib import Path

warnings.filterwarnings('ignore')

# Set style
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (12, 6)

# Set paths
data_path = Path('../../data/Customers.csv')
results_path = Path('../../results')
results_path.mkdir(exist_ok=True)

# Load data
df = pd.read_csv(data_path)

def descriptive_statistics():
    """Calculate descriptive statistics"""
    print("=" * 50)
    print("DESCRIPTIVE STATISTICS")
    print("=" * 50)
    
    numerical_cols = df.select_dtypes(include=[np.number]).columns.tolist()
    
    print("\n=== Descriptive Statistics for Numerical Variables ===")
    print(df[numerical_cols].describe())
    
    # Additional descriptive statistics
    print("\n=== Additional Descriptive Statistics ===")
    desc_stats = pd.DataFrame({
        'Mean': df[numerical_cols].mean(),
        'Median': df[numerical_cols].median(),
        'Std Dev': df[numerical_cols].std(),
        'Variance': df[numerical_cols].var(),
        'Min': df[numerical_cols].min(),
        'Max': df[numerical_cols].max(),
        'Range': df[numerical_cols].max() - df[numerical_cols].min(),
        'Q1': df[numerical_cols].quantile(0.25),
        'Q3': df[numerical_cols].quantile(0.75),
        'IQR': df[numerical_cols].quantile(0.75) - df[numerical_cols].quantile(0.25),
        'Skewness': df[numerical_cols].skew(),
        'Kurtosis': df[numerical_cols].kurtosis()
    })
    print(desc_stats.T)
    
    # Categorical variables
    categorical_cols = df.select_dtypes(include=['object']).columns.tolist()
    print("\n=== Descriptive Statistics for Categorical Variables ===")
    for col in categorical_cols:
        print(f"\n{col}:")
        print(f"  Count: {df[col].count()}")
        print(f"  Unique values: {df[col].nunique()}")
        print(f"  Mode: {df[col].mode()[0] if len(df[col].mode()) > 0 else 'N/A'}")
        print(f"  Mode frequency: {df[col].value_counts().iloc[0] if len(df[col].value_counts()) > 0 else 'N/A'}")

def inferential_statistics():
    """Perform inferential statistics"""
    print("\n" + "=" * 50)
    print("INFERENTIAL STATISTICS")
    print("=" * 50)
    
    numerical_cols = df.select_dtypes(include=[np.number]).columns.tolist()
    
    # Confidence intervals
    print("\n=== Confidence Intervals (95%) ===")
    for col in numerical_cols:
        mean = df[col].mean()
        std = df[col].std()
        n = len(df[col])
        
        margin_error = 1.96 * (std / np.sqrt(n))
        ci_lower = mean - margin_error
        ci_upper = mean + margin_error
        
        print(f"\n{col}:")
        print(f"  Mean: {mean:.2f}")
        print(f"  95% CI: [{ci_lower:.2f}, {ci_upper:.2f}]")
    
    # Normality tests
    print("\n=== Normality Tests ===")
    alpha = 0.05
    for col in numerical_cols:
        data = df[col].dropna()
        stat, p_value = normaltest(data)
        
        print(f"\n{col}:")
        print(f"  D'Agostino-Pearson test:")
        print(f"    Statistic: {stat:.4f}")
        print(f"    p-value: {p_value:.4f}")
        
        if p_value > alpha:
            print(f"    Result: Data appears to be normally distributed (p > {alpha})")
        else:
            print(f"    Result: Data does not appear to be normally distributed (p <= {alpha})")
    
    # Chi-square test
    print("\n=== Chi-Square Test for Independence ===")
    contingency_table = pd.crosstab(df['Country'], df['City'])
    chi2, p_value, dof, expected = chi2_contingency(contingency_table)
    
    print(f"\nChi-square statistic: {chi2:.4f}")
    print(f"Degrees of freedom: {dof}")
    print(f"p-value: {p_value:.4f}")
    
    alpha = 0.05
    if p_value < alpha:
        print(f"\nResult: Reject null hypothesis. Country and City are not independent (p < {alpha})")
    else:
        print(f"\nResult: Fail to reject null hypothesis. Country and City may be independent (p >= {alpha})")

def exploratory_statistics():
    """Perform exploratory statistical analysis"""
    print("\n" + "=" * 50)
    print("EXPLORATORY STATISTICAL ANALYSIS")
    print("=" * 50)
    
    numerical_cols = df.select_dtypes(include=[np.number]).columns.tolist()
    
    # Distribution analysis
    print("\n=== Distribution Analysis ===")
    for col in numerical_cols:
        fig, axes = plt.subplots(1, 2, figsize=(14, 5))
        
        # Histogram
        axes[0].hist(df[col], bins=20, edgecolor='black', alpha=0.7, color='steelblue')
        axes[0].set_xlabel(col)
        axes[0].set_ylabel('Frequency')
        axes[0].set_title(f'Distribution of {col}')
        axes[0].grid(True, alpha=0.3)
        
        # Box plot
        axes[1].boxplot(df[col], vert=True)
        axes[1].set_ylabel(col)
        axes[1].set_title(f'Box Plot of {col}')
        axes[1].grid(True, alpha=0.3)
        
        plt.tight_layout()
        plt.savefig(results_path / f'{col}_distribution.png', dpi=300, bbox_inches='tight')
        plt.close()
    
    # Outlier detection
    print("\n=== Outlier Detection (IQR Method) ===")
    for col in numerical_cols:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR
        
        outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)][col]
        
        print(f"\n{col}:")
        print(f"  Lower bound: {lower_bound:.2f}")
        print(f"  Upper bound: {upper_bound:.2f}")
        print(f"  Number of outliers: {len(outliers)}")
        if len(outliers) > 0:
            print(f"  Outlier values: {outliers.tolist()}")
    
    # Central tendency and dispersion
    print("\n=== Central Tendency and Dispersion Measures ===")
    measures = pd.DataFrame({
        'Mean': df[numerical_cols].mean(),
        'Median': df[numerical_cols].median(),
        'Std Dev': df[numerical_cols].std(),
        'Variance': df[numerical_cols].var(),
        'CV (%)': (df[numerical_cols].std() / df[numerical_cols].mean() * 100)
    })
    print(measures)
    
    print("\n✓ Statistical analysis visualizations saved to results/ directory")

def main():
    """Main function"""
    descriptive_statistics()
    inferential_statistics()
    exploratory_statistics()
    
    print("\n" + "=" * 50)
    print("STATISTICAL ANALYSIS SUMMARY")
    print("=" * 50)
    print("\n1. Descriptive Statistics:")
    print("   • Calculated measures of central tendency (mean, median, mode)")
    print("   • Calculated measures of dispersion (std dev, variance, IQR)")
    print("   • Analyzed distribution characteristics (skewness, kurtosis)")
    print("\n2. Inferential Statistics:")
    print("   • Calculated 95% confidence intervals")
    print("   • Performed normality tests")
    print("   • Conducted chi-square tests for independence")
    print("\n3. Exploratory Statistics:")
    print("   • Analyzed distributions")
    print("   • Detected outliers")
    print("   • Examined central tendency and dispersion")

if __name__ == "__main__":
    main()

File: Customer Data Analysis and Insights Project/scripts/r/statistical_analysis.R
# Statistical Analysis Script
# Descriptive, Inferential, and Exploratory Statistical Analysis

# Load libraries
library(tidyverse)
library(ggplot2)
library(dplyr)
library(corrplot)

# Set paths
data_path <- "../../data/Customers.csv"
results_path <- "../../results"

# Create results directory if it doesn't exist
if (!dir.exists(results_path)) {
  dir.create(results_path, recursive = TRUE)
}

# Load data
df <- read.csv(data_path, stringsAsFactors = FALSE)

cat("==================================================\n")
cat("STATISTICAL ANALYSIS\n")
cat("==================================================\n\n")

# 1. Descriptive Statistics
cat("=== DESCRIPTIVE STATISTICS ===\n\n")

# Numerical variables
numerical_cols <- sapply(df, is.numeric)
if (sum(numerical_cols) > 0) {
  cat("=== Numerical Variables Descriptive Statistics ===\n")
  print(summary(df[, numerical_cols, drop = FALSE]))
  
  # Additional descriptive statistics
  cat("\n=== Additional Descriptive Statistics ===\n")
  desc_stats <- df[, numerical_cols, drop = FALSE] %>%
    summarise_all(list(
      Mean = ~mean(., na.rm = TRUE),
      Median = ~median(., na.rm = TRUE),
      StdDev = ~sd(., na.rm = TRUE),
      Variance = ~var(., na.rm = TRUE),
      Min = ~min(., na.rm = TRUE),
      Max = ~max(., na.rm = TRUE),
      Q1 = ~quantile(., 0.25, na.rm = TRUE),
      Q3 = ~quantile(., 0.75, na.rm = TRUE),
      IQR = ~IQR(., na.rm = TRUE)
    ))
  print(desc_stats)
} else {
  cat("No numerical columns found.\n")
}

# Categorical variables
cat("\n=== Categorical Variables Descriptive Statistics ===\n")
categorical_cols <- sapply(df, is.character) | sapply(df, is.factor)
for (col in names(df)[categorical_cols]) {
  cat("\n", col, ":\n", sep = "")
  cat("  Count:", nrow(df), "\n")
  cat("  Unique values:", n_distinct(df[[col]]), "\n")
  cat("  Mode:", names(sort(table(df[[col]]), decreasing = TRUE)[1]), "\n")
  cat("  Mode frequency:", max(table(df[[col]])), "\n")
  cat("  Mode percentage:", round((max(table(df[[col]])) / nrow(df)) * 100, 2), "%\n")
}

# 2. Inferential Statistics
cat("\n==================================================\n")
cat("INFERENTIAL STATISTICS\n")
cat("==================================================\n\n")

# Confidence Intervals
if (sum(numerical_cols) > 0) {
  cat("=== Confidence Intervals (95%) ===\n")
  for (col in names(df)[numerical_cols]) {
    data <- df[[col]][!is.na(df[[col]])]
    mean_val <- mean(data)
    std_val <- sd(data)
    n <- length(data)
    
    margin_error <- 1.96 * (std_val / sqrt(n))
    ci_lower <- mean_val - margin_error
    ci_upper <- mean_val + margin_error
    
    cat("\n", col, ":\n", sep = "")
    cat("  Mean:", round(mean_val, 2), "\n")
    cat("  95% CI: [", round(ci_lower, 2), ", ", round(ci_upper, 2), "]\n", sep = "")
  }
}

# Normality Tests
if (sum(numerical_cols) > 0) {
  cat("\n=== Normality Tests ===\n")
  alpha <- 0.05
  for (col in names(df)[numerical_cols]) {
    data <- df[[col]][!is.na(df[[col]])]
    
    # Shapiro-Wilk test (for small samples)
    if (length(data) <= 5000) {
      shapiro_test <- shapiro.test(data)
      cat("\n", col, ":\n", sep = "")
      cat("  Shapiro-Wilk test:\n")
      cat("    Statistic:", round(shapiro_test$statistic, 4), "\n")
      cat("    p-value:", round(shapiro_test$p.value, 4), "\n")
      
      if (shapiro_test$p.value > alpha) {
        cat("    Result: Data appears to be normally distributed (p > ", alpha, ")\n", sep = "")
      } else {
        cat("    Result: Data does not appear to be normally distributed (p <= ", alpha, ")\n", sep = "")
      }
    }
  }
}

# Chi-Square Test for Independence
cat("\n=== Chi-Square Test for Independence ===\n")
if ("Country" %in% colnames(df) && "City" %in% colnames(df)) {
  contingency_table <- table(df$Country, df$City)
  chi2_test <- chisq.test(contingency_table)
  
  cat("\nChi-square statistic:", round(chi2_test$statistic, 4), "\n")
  cat("Degrees of freedom:", chi2_test$parameter, "\n")
  cat("p-value:", round(chi2_test$p.value, 4), "\n")
  
  alpha <- 0.05
  if (chi2_test$p.value < alpha) {
    cat("\nResult: Reject null hypothesis. Country and City are not independent (p < ", alpha, ")\n", sep = "")
  } else {
    cat("\nResult: Fail to reject null hypothesis. Country and City may be independent (p >= ", alpha, ")\n", sep = "")
  }
}

# 3. Exploratory Statistical Analysis
cat("\n==================================================\n")
cat("EXPLORATORY STATISTICAL ANALYSIS\n")
cat("==================================================\n\n")

# Distribution Analysis
if (sum(numerical_cols) > 0) {
  cat("=== Distribution Analysis ===\n")
  for (col in names(df)[numerical_cols]) {
    # Histogram
    png(file = file.path(results_path, paste0(col, "_histogram.png")), 
        width = 1400, height = 600, res = 300)
    hist(df[[col]], breaks = 20, main = paste("Distribution of", col),
         xlab = col, ylab = "Frequency", col = "steelblue", border = "black")
    dev.off()
    
    # Box plot
    png(file = file.path(results_path, paste0(col, "_boxplot.png")), 
        width = 800, height = 600, res = 300)
    boxplot(df[[col]], main = paste("Box Plot of", col), ylab = col, col = "lightblue")
    dev.off()
    
    cat("✓ Plots saved for", col, "\n")
  }
}

# Outlier Detection
if (sum(numerical_cols) > 0) {
  cat("\n=== Outlier Detection (IQR Method) ===\n")
  for (col in names(df)[numerical_cols]) {
    Q1 <- quantile(df[[col]], 0.25, na.rm = TRUE)
    Q3 <- quantile(df[[col]], 0.75, na.rm = TRUE)
    IQR_val <- Q3 - Q1
    
    lower_bound <- Q1 - 1.5 * IQR_val
    upper_bound <- Q3 + 1.5 * IQR_val
    
    outliers <- df[[col]][df[[col]] < lower_bound | df[[col]] > upper_bound]
    
    cat("\n", col, ":\n", sep = "")
    cat("  Lower bound:", round(lower_bound, 2), "\n")
    cat("  Upper bound:", round(upper_bound, 2), "\n")
    cat("  Number of outliers:", length(outliers), "\n")
    if (length(outliers) > 0) {
      cat("  Outlier values:", paste(outliers, collapse = ", "), "\n")
    }
  }
}

# Central Tendency and Dispersion
if (sum(numerical_cols) > 0) {
  cat("\n=== Central Tendency and Dispersion Measures ===\n")
  measures <- df[, numerical_cols, drop = FALSE] %>%
    summarise_all(list(
      Mean = ~mean(., na.rm = TRUE),
      Median = ~median(., na.rm = TRUE),
      StdDev = ~sd(., na.rm = TRUE),
      Variance = ~var(., na.rm = TRUE),
      CV = ~(sd(., na.rm = TRUE) / mean(., na.rm = TRUE)) * 100
    ))
  print(measures)
}

cat("\n==================================================\n")
cat("STATISTICAL ANALYSIS SUMMARY\n")
cat("==================================================\n")
cat("\n1. Descriptive Statistics:\n")
cat("   • Calculated measures of central tendency (mean, median, mode)\n")
cat("   • Calculated measures of dispersion (std dev, variance, IQR)\n")
cat("   • Analyzed distribution characteristics\n")
cat("\n2. Inferential Statistics:\n")
cat("   • Calculated 95% confidence intervals\n")
cat("   • Performed normality tests\n")
cat("   • Conducted chi-square tests for independence\n")
cat("\n3. Exploratory Statistics:\n")
cat("   • Analyzed distributions\n")
cat("   • Detected outliers\n")
cat("   • Examined central tendency and dispersion\n")

cat("\n✓ Statistical analysis completed successfully!\n")

